crypto/crypto/src/multi_ed25519.rs

```diff
@@ -0,0 +1,558 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides an API for the accountable threshold multi-sig PureEdDSA signature scheme
+//! over the ed25519 twisted Edwards curve as defined in [RFC8032](https://tools.ietf.org/html/rfc8032).
+//!
+//! Signature verification also checks and rejects non-canonical signatures.
+
+use crate::ed25519::{
+    Ed25519PrivateKey, Ed25519PublicKey, Ed25519Signature, ED25519_PRIVATE_KEY_LENGTH,
+    ED25519_PUBLIC_KEY_LENGTH, ED25519_SIGNATURE_LENGTH,
+};
+use crate::{traits::*, HashValue};
+use anyhow::{anyhow, Result};
+use bit_vec::BitVec;
+use core::convert::TryFrom;
+use libra_crypto_derive::{DeserializeKey, SerializeKey, SilentDebug, SilentDisplay};
+use once_cell::sync::Lazy;
+use rand::Rng;
+use std::fmt;
+
+const MAX_NUM_OF_KEYS: usize = 32;
+const BITVEC_NUM_OF_BYTES: usize = MAX_NUM_OF_KEYS / 8;
+static BITVEC_FIRST_KEY_SET: Lazy<BitVec> =
+    Lazy::new(|| BitVec::from_bytes(&[0b1000_0000, 0b0000_0000, 0b0000_0000, 0b0000_0000]));
+
+/// Vector of private keys in the multi-key Ed25519 structure along with the threshold.
+#[derive(DeserializeKey, Eq, PartialEq, SilentDisplay, SilentDebug, SerializeKey)]
+pub struct MultiEd25519PrivateKey {
+    private_keys: Vec<Ed25519PrivateKey>,
+    threshold: u8,
+}
+
+#[cfg(feature = "assert-private-keys-not-cloneable")]
+static_assertions::assert_not_impl_any!(MultiEd25519PrivateKey: Clone);
+
+/// Vector of public keys in the multi-key Ed25519 structure along with the threshold.
+#[derive(Clone, DeserializeKey, Eq, PartialEq, SerializeKey)]
+pub struct MultiEd25519PublicKey {
```

Comment:
This can be done in a follow-up, but we may need to change this to

```
pub struct MultisigPolicy<PublicKey: VerifyingKey> {
    public_keys: Vec<PublicKey>,
    threshold: u8,
}
```
(and similar for the private key case)

Most of the  uses of public keys in the codebase seems to use `PublicKey: VerifyingKey` rather than committing to a particular implementation... I  was just trying to use this new public key type in account_address.rs and ended up needing to use the more generic form.

Reply:
As discussed, `MultiEd25519PublicKey` is a `VerifyingKey` which can simplify issues re generics, so we've agreed w/ Sam that it can stay without modification and we just handle it as any other crypto scheme.